### DISCUSSION TITLE
How does wiliwili detect system theme in ?

### DISCUSSION THREAD

**Discussion by yuioto** [xfangfang/wiliwili#177] [Q&A] [answered]
# wiliwili 的自动暗黑模式在我的系统上不起作用。
我猜想可能是wiliwili读取的环境变量或其它设置没有被我设置，我应该怎么做才能让wiliwili知道我在何种主题？

## system info

`ArchLinux` Install for aur `wiliwili-git` ver `1.1.1.r0.g7541334-1`

`Desktop Environment` is `Xfce4`

Use `darkman` (install for aur darkman)

## darkman 切换脚本
我通过让darkman执行脚本，借助脚本设置主题。下面是这些脚本的实现。

### Dark
``` shell
theme="Arc-Dark"

xfconf-query -c xsettings -p /Net/ThemeName -s "$theme"
xfconf-query -c xfwm4 -p /general/theme -s "$theme"

gsettings set org.gnome.desktop.interface gtk-theme "$theme"
```
### Light
``` shell
theme="Quartz Green"

xfconf-query -c xsettings -p /Net/ThemeName -s "$theme"
xfconf-query -c xfwm4 -p /general/theme -s "$theme"

gsettings set org.gnome.desktop.interface gtk-theme "$theme"
```

## darkman 目前的工作状况
目前这个脚本对GTK和Tauri都能工作的很好。~~暂时没有写QT程序的部分。~~
在进行了一些设置，现在QT主题表现的和GTK主题一样。
https://wiki.archlinux.org/title/Uniform_look_for_Qt_and_GTK_applications#QGtkStyle

## wiliwili 是怎么识别当前系统主题明暗风格的？
如果我能知道wiliwili如何实现的自动切换主题，我就能通过脚本让wiliwili在我的系统上正常工作。

**Comment by xfangfang** [comment, marked as answer]
wiliwili目前还不支持运行时检测系统主题。

1. 在窗口初始化时

读取环境变量：BOREALIS_THEME。若未设定，PC平台在 macOS 和 UWP 下自动获取系统主题，其他系统默认为白色主题

https://github.com/xfangfang/borealis/blob/a3e600e5e6f5b0e1bf7546fa47ff075296cf4357/library/lib/platforms/desktop/desktop_platform.cpp#L370-L401

BOREALIS_THEME 可能为的值是：DARK / LIGHT

2. 在初始化之后
读取配置目录，如果有应用内强制设定的主题，还会覆盖前面的设置

https://github.com/xfangfang/wiliwili/blob/7541334784c96991d6d391a5acda5410a875f1af/wiliwili/source/utils/config_helper.cpp#L485-L494

**Comment by yuioto** [comment]
This can be temporarily solved using a script that gets the theme color scheme at startup and sets the environment variable.

Move wiliwili to share dir.

```shell
#!/bin/env bash
#
# org.freedesktop.appearance color-scheme
#
# Indicates the system's preferred color scheme.
# Supported values are:
#
#   0: No preference
#   1: Prefer dark appearance
#   2: Prefer light appearance
#
# Unknown values should be treated as 0 (no preference).

wiliwili_path=/usr/share/wiliwili/wiliwili

scheme=$(
  gdbus call --session --timeout=1000 \
             --dest=org.freedesktop.portal.Desktop \
             --object-path /org/freedesktop/portal/desktop \
             --method org.freedesktop.portal.Settings.Read org.freedesktop.appearance color-scheme
)

wiliwili_boot_dark () {
	BOREALIS_THEME=DARK $wiliwili_path
}

wiliwili_boot_light () {
	BOREALIS_THEME=LIGHT $wiliwili_path
}

wiliwili_boot_default () {
	wiliwili_boot_light
}

case $scheme in
  ( '(<<uint32 1>>,)' ) wiliwili_boot_dark;;
  ( '(<<uint32 2>>,)' ) wiliwili_boot_light;;
  ( *                 ) wiliwili_boot_default;;
esac

```

Copy from: 
https://aur.archlinux.org/cgit/aur.git/tree/wiliwili.sh?h=wiliwili-git&id=e6fb90e5a797d56b1d13c6adcbe15fba9f5a31ff
